src/Groups/matrices/linear_centralizer.jl

```diff
@@ -0,0 +1,385 @@
+# generators for GL(n,q) and SL(n,q) are described in:  Taylor, D. E., Pairs of Generators for Matrix Groups. I, The Cayley Bulletin, 3 (1987)
+
+# returns as matrices
+function _gens_for_GL(n::Int, F::Ring)
+   n !=1 || return [matrix(F,1,1,[primitive_root(F)])]
+   h1 = identity_matrix(F,n)
+   h2 = zero_matrix(F,n,n)
+   if order(F)==2
+      h1[1,2] = 1
+   else
+      h1[1,1] = primitive_root(F)
+      h2[1,1] = -1
+   end
+   h2[1,n] = 1
+   for i in 1:n-1 h2[i+1,i] = -1 end
+   return h1,h2
+end
+
+# returns as matrices
+# does the same as above with F = F[x]/(f), but every entry is replaced by a diagonal join of D corresponding blocks
+# ASSUMPTION: deg(f) > 1
+function _gens_for_GL_matrix(f::PolyElem, n::Int, F::Ring; D=1)
+   C = companion_matrix(f)
+   CP = evaluate(_centralizer(f),C)            # matrix of maximal order in the centralizer of the companion matrix
+
+   if n==1 return [diagonal_join([CP for i in 1:D])] end
+   h1 = identity_matrix(F,n*degree(f)*D)
+   insert_block!(h1,diagonal_join([CP for i in 1:D]),1,1)
+   h2 = zero_matrix(F,n*degree(f)*D,n*degree(f)*D)
+   for i in 1:(n-1)*degree(f)*D h2[i+degree(f)*D,i]=-1 end
+   for i in 1:degree(f)*D
+      h2[i,i]=-1
+      h2[i,i+(n-1)*degree(f)*D]=1
+   end
+   return h1,h2
+end
+
+# generators for centralizer of unipotent elements are described in: Giovanni De Franceschi, Centralizers and conjugacy classes in finite classical groups, arXiv:2008.12651
+
+# V = vector of integers of the dimensions of Jordan blocks
+# return the generators for the centralizers of the diagonal join of unipotent Jordan blocks of dimensions V
+# assumes V is sorted (e.g. [1,1,1,2,3,3])
+function _centr_unipotent(F::Ring, V::AbstractVector{Int}; isSL=false)
+   n = sum(V)
+   _lambda = gen(F)  # yes, gen(F) is correct; we don't need a primitive element in this case
+
+   # L = multiset(V)
+   L=[[V[1],1]]
+   for i in 2:length(V)
+      if V[i]==L[length(L)][1]
+         L[length(L)][2] += 1
+      else
+         push!(L, [V[i],1])
+      end
+   end
+   listgens = MatElem[]
+
+   # generators for GL + internal diagonal blocks
+   pos=1
+   for l in L
+      v_g = isSL ? _gens_for_SL(l[2],F) : _gens_for_GL(l[2],F)
+      for x in v_g
+         z = block_matrix(l[2],l[2],[x[i,j]*identity_matrix(F,l[1]) for i in 1:l[2] for j in 1:l[2]])
+         z = insert_block(identity_matrix(F,n),z,pos,pos)
+         push!(listgens,z)
+      end
+      if l[1]>1
+         for i in 1:l[1]-1, j in 1:degree(F)
+            z = identity_matrix(F,l[1])
+            for k in 1:l[1]-i z[k,i+k]=_lambda^j end
+            z = insert_block(identity_matrix(F,n),z,pos,pos)
+            push!(listgens,z)
+         end
+      end
+      pos += l[1]*l[2]
+   end
+
+   # external diagonal blocks
+   pos=1
+   for i in 1:length(L)-1
+      pos += L[i][1]*L[i][2]
+      # block above diagonal
+      z = identity_matrix(F,n)
+      for j in 1:L[i][1] z[pos-L[i][1]+j-1,pos+L[i+1][1]-L[i][1]+j-1]=1 end
```

Comment:
This is really difficult to read. I'd not squish it into a single line. Also, perhaps use more bespoke names?

Maybe this (please double check I didn't mess it up)
```suggestion
      for j in 1:L[i][1]
         idx = pos-L[i][1]+j-1
         z[idx,idx+L[i+1][1]] = 1
      end
```
or
```suggestion
      idx = pos-L[i][1]-1
      for j in 1:L[i][1]
         idx += 1
         z[idx,idx+L[i+1][1]] = 1
      end
```

Reply:
Corrected in all the places it occurs.